lib/firenest/topology/erlang.ex

```diff
@@ -133,112 +126,316 @@ defmodule Firenest.Topology.Erlang do
+          state
+      end
+
+    {:noreply, state}
+  end
+
+  def handle_info({:monitor_down, node, id, name, monitor_ref}, state) do
+    %{nodes: nodes, local_names: local_names} = state
+
+    state =
+      case nodes do
+        # We know this node and therefore we must know this name-monitor pair.
+        %{^node => {^id, node_ref, remote_names}} ->
+          ^monitor_ref = Map.fetch!(remote_names, name)
+          local_monitor_down(local_names, node, id, name)
+          put_in(state.nodes[node], {id, node_ref, Map.delete(remote_names, name)})
+
+        # We either have a mismatched or an unknown ID because the
+        # PONG message has not been processed yet.
+        _ ->
+          state
+      end
+
     {:noreply, state}
   end
 
+  # This message comes from :net_kernel.monitor_nodes/2. Note it is not
+  # guaranteed that we will receive a nodedown before a nodeup with the
+  # same name. More info: http://erlang.org/pipermail/erlang-questions/2016-November/090795.html
+  def handle_info({:nodeup, node, _}, state) do
+    %{topology: topology, id: id, monitors: monitors} = state
+    ping(node, topology, id, Map.to_list(monitors))
+    {:noreply, state}
+  end
+
+  # Sent by :net_kernel.monitor_nodes/2 on node down. We don't worry
+  # about it because we already monitor the topology process and the
+  # monitor message is guaranteed to be delivered before nodedown.
   def handle_info({:nodedown, _, _}, state) do
     {:noreply, state}
   end
 
-  def handle_info({:ping, other_id, pid}, %{id: id} = state) do
-    pong(pid, id)
-    {:noreply, add_node(state, other_id, pid)}
+  # If two nodes come up at the same time, ping may be received twice.
+  # So we need to make sure to handle two pings/pongs.
+  def handle_info({:ping, other_id, pid, monitors}, %{id: id} = state) do
+    pong(pid, id, Map.to_list(state.monitors))
+    {:noreply, add_node(state, other_id, pid, monitors)}
   end
 
-  def handle_info({:pong, other_id, pid}, state) do
-    {:noreply, add_node(state, other_id, pid)}
+  def handle_info({:pong, other_id, pid, monitors}, state) do
+    {:noreply, add_node(state, other_id, pid, monitors)}
   end
 
   def handle_info({:DOWN, ref, _, pid, _}, state) when Kernel.node(pid) != Kernel.node() do
     {:noreply, delete_node(state, pid, ref)}
   end
 
-  def handle_info({:DOWN, ref, _, _, _}, state) do
-    {_, state} = pop_in(state.subscribers[ref])
-    {:noreply, state}
-  end
+  ## Helpers
 
-  defp ping(node, topology, id) do
-    send({topology, node}, {:ping, id, self()})
+  defp ping(node, topology, id, monitors) when is_list(monitors) do
+    send({topology, node}, {:ping, id, self(), monitors})
   end
 
-  defp pong(pid, id) do
-    send(pid, {:pong, id, self()})
+  defp pong(pid, id, monitors) when is_list(monitors) do
+    send(pid, {:pong, id, self(), monitors})
   end
 
   defp id() do
-    {:crypto.strong_rand_bytes(12), System.system_time()}
+    {:crypto.strong_rand_bytes(4), System.system_time()}
```

Comment:
Would it make sense to make the id a binary instead of a tuple? Also, if the `system_time` part was the first element it would be sortable - maybe that would be beneficial in some circumstances?

Reply:
This is a completely opaque term so I don't want to say it is ordered because other topologies may not give the same guarantee.

Reply:
If you used `{Kernel.node(), System.unique_integer([:monotonic])`? You could guarantee that ids produced by the same node are both unique and are ordered, perhaps that would help in the case of the ping/pong issue @fishcakez described, as a node restarting would generate a new id which is strictly larger than the previous id, so you could correlate any messages to the "instance" of that node which generated it

Reply:
The problem with ping/pong is strictly when network partitions occur and the ids don't change on either node.  It is about ensuring the protocol is serializable so we can provide guarantees of `monitor_up` and `monitor_down`. Any unique "enough" id would not suffer the issue as ids would be different so messages couldn't be confused.

Reply:
Ah right, should've caught that. What about generating an "id" for each event using `unique_integer`, and tracking the most recent id received from a node, so that you could effectively ignore messages which are in the past relative to whatever has already been received, e.g. if a node flaps, it is theoretically possible to get a nodeup before the nodedown is received, but if the ping/pong between the nodes has a unique id associated with it generated with `unique_integer`, there is an absolute ordering of those events, and if a pong is received which causally occurred before the nodedown, but is received _after_ the nodedown, then it is clear the pong can be ignored, as it arrived out of order relative to the actual causal timeline and has a unique integer "less than" the nodedown. I find it hard to trust my intuitions when it comes to this stuff though, so I hope I'm not missing something obvious.

Reply:
The pid -> pid message ordering is already guaranteed to be in order. Therefore if a `unique_integer([:monotonic])` is added to each event the next event will always have a larger id so we would always accept all messages and nor gain from the strictly monotonically increasing id.

The `nodeup`/`nodedown` flapping issue is already avoided by ignoring `nodedown` and using a process monitor on the remote pid.

The difficulty is more in that there is no guarantee of delivery and that messages from prior handshake (not just `pong` but also `monitor_up` and `monitor_down`) can arrive late.

Jose and I discussed this offline this morning and we came up with two ways to handle it:

* Require all messages to use the monitor reference from the ping, then when a new ping is sent (after a `nodeup`) the new messages must have a new monitor reference in them or be ignored.

* Use a (lamport) logical clock on each node that increments every time that node's state changes (i.e. `monitor_up` or `monitor_down`), include the counter in every message it sends, if the receiving node misses a state change (counter increments by more 1) then treat it as disconnect between it and the sending node and send a new `ping`. If we don't fake the disconnect then we lose serializability as we can get impossible ordering such as two `monitor_up` for the same node/name without a `monitor_down` in between. This also allows some optimizations like pairing "old" pongs with newer pings.

We decided to go for the second approach, and it should be easier to translate to other adapters.